### ISSUE TITLE
Documentation change: easy_install to pip

### ISSUE THREAD

**Issue by stevepiercy** [Pylons/pyramid#2104]
Update the documentation to use pip instead of easy_install.

Work should be pushed to the branch [docs/easy-install-to-pip.2104](https://github.com/Pylons/pyramid/tree/docs/easy-install-to-pip.2104).

Branch `docs/easy-install-to-pip.2104` will eventually be merged to `master` upon completion of the items listed below.

Add omitted items to the lists.

# List of related issues and PRs.
- https://github.com/Pylons/pyramid/issues/575
- https://github.com/Pylons/pyramid/issues/1454
- https://github.com/Pylons/pyramid/issues/121
- https://github.com/Pylons/pyramid/pull/2024

# List of documentation and official tutorials
(Pretty much every step for tutorials needs updating, so only the indices are shown for brevity.)
- [x] [Glossary](http://docs.pylonsproject.org/projects/pyramid/en/master/glossary.html)
- [x] [Quick Tour](http://docs.pylonsproject.org/projects/pyramid/en/master/quick_tour.html)
- [x] [Quick Tutorial](http://docs.pylonsproject.org/projects/pyramid/en/master/quick_tutorial/index.html)
- [x] [SQLAlchemy + URL Dispatch Wiki Tutorial](http://docs.pylonsproject.org/projects/pyramid/en/master/tutorials/wiki2/index.html)
- [x] [ZODB + Traversal Wiki Tutorial](http://docs.pylonsproject.org/projects/pyramid/en/master/tutorials/wiki/index.html)
- [x] [Running a Pyramid Application under mod_wsgi](http://docs.pylonsproject.org/projects/pyramid/en/master/tutorials/modwsgi/index.html)
- [x] narr/commandline.rst
- [x] narr/extending.rst
- [x] narr/installation.html
- [x] narr/project.rst
- [x] narr/testing.rst

**Comment by mmerickel** [comment]
A couple notes on how to do this:

1. `$VENV/bin/python setup.py develop` is now `$VENV/bin/pip install -e .` (the `.` on the end is important).
2. `$VENV/bin/python setup.py install` is now `$VENV/bin/pip install .

One big issue when updating to pip is that `python setup.py test` is no longer a great option for running tests. Specifically because the `tests_require` option will be installed using setuptools/easy_install instead of pip. We need to come up with a common pattern for this.

Two options (please chime in with some alternatives) I'm aware of that would work alright:

1. Add a `testing` extras to `setup.py` and show people how to install them with `$VENV/bin/pip install -e .[testing]`. They can then continue running their tests with `$VENV/bin/python setup.py test`.

   Pros: The `testing` extras is a good pattern.
   Cons: The `python setup.py test` will try to install (using setuptools instead of pip) any missing dependencies if someone forgot to run `-e .[testing]`.

2. Use `tox` to define the dependencies in a `tox.ini` and recommend people run `tox` to start the tests.

   Pros: Tox is great, and encapsulated, and uses `pip` by default.
   Cons: Most non-library-authors are probably less excited about using tox in their projects and is a larger change from what they were doing before.

3. Add a `testing` extras to `setup.py` and show people `$VENV/bin/pip install -e .[testing]` and advocate running the tests with `$VENV/bin/nosetests`.

   Pros: This is probably the best way to go. It removes `setup.py` from the equation entirely as a top-level executable.
   Cons: Requires us to define a test runner versus using setuptools' discovery mechanism. It's possible to avoid dependencies on nose or py.test by using [python's default unittest runner](https://docs.python.org/2/library/unittest.html#test-discovery) but it's a little kludgey and only available in 2.7+ (not 2.6). This means we'd want to add a dependency on `nose` or `py.test` (my personal choice). However this may not sit will with people.

Let's talk about it and come to some conclusions so we can migrate everything consistently.

**Comment by pauleveritt** [comment]
Geez, the number of things I don’t know. .[testing] ? Really? I’ve read all this stuff numerous times.

My two cents, for what it’s worth:

- The switch to pip is about recognizing that our docs need to appeal to people using mainstream Python choices

- Stated differently, if they expect pip, and we give easy_install, we’re the weird ones, not them (despite our valid protests)

- I feel the same about my personal switch to jinja2 for all teaching material

That said, I don’t think there’s a strong majority that is doing this a certain way. I personally don’t see many docs out there showing test running via setup.py. I think py.test is perhaps a cooler-kid on the block than nose, but not overwhelmingly (though it does seem more active.)

So I vote for (3) with py.test, caveat that it means fixing a lot of docs away from nose.

Regarding best practices, any implications on how Travis works? Does it sniff at setup.py to discover what to run, or does it just do what you tell it in the YAML file?

—Paul

**Comment by mgrbyte** [comment]
Not being a regular contributor, not sure my vote counts....
but my 2p (British money...) is same Paul, I'd (option 3 w/py.test instead of nose)

One caveat - when using the `zsh` shell, you need to quote the extras, i.e:
```
pip install ---editable ".[testing]" 
```

**Comment by blaflamme** [comment]
I'm for #3 too.

**Comment by mmerickel** [comment]
@dstufft I'd love to hear your thoughts on test runners and `tests_require` usage in this brave new world. My impression is that `tests_require` and `test_suite` should no longer be used and focus on using `setup.py` as the metadata description/build system rather than the runner.

**Comment by dstufft** [comment]
I normally do a combination of all three, I define a ``test`` extra and I have the same list being sent into both that extra and the ``tests_require`` keyword. I tend to override the ``test`` command to use py.test as well and then I pretty much always just run the tests using tox :).

That being said, the story around a standard interface for running tests (other than a sort of defacto stand off between tox and ``setup.py test``) isn't great right now. It's mostly about picking tradeoffs right now :(

**Comment by bertjwregeer** [comment]
I tend to use `py.test` for all of my testing, run through `tox`. This way I can build documentation and everything else along those lines automatically as well.

**Comment by jvanasco** [comment]
I would just really love it if package maintainers would simply make a top-level `TESTS.txt` that documents how the tests are organized and invoked, and gives pointers on how to customize testing for feature development or bugfixing.

In the past 12 months, I've contributed patches + tests to at least that many python projects.  No two projects have taken the same approach to tests, and many people use "bleeding edge" test harnesses that look like they should be invoked one way, but really want to be used another (ie, if you run via `setup.py` then suites a+b+c will run... if you run via something else d+e+f will run, and if you run via a 3rd option a+b+c+d+e+f will run).  I've had to learn way more testing frameworks and random implementations than I'd like.

So whatever you choose, please please please just make a TESTS.txt file that at least says:

* We use ___ for tests.
* You can invoke the full testing suite by doing this _____
* If you'd like to run a subset of tests to isolate an issue or fix, you should do _______
* When writing new tests, we prefer ______

and give a few examples.

If you can convey that information (and maybe a few other things) and keep it up-to-date, many people will be exceptionally happy.

**Comment by tseaver** [comment]
@jvanasco maybe you mean [this](https://github.com/Pylons/pyramid/blob/e70e1e2b851a8387f0351a19b06339ed69693ef6/HACKING.txt#L161-L202)?

**Comment by bertjwregeer** [comment]
@tseaver :+1:

**Comment by jvanasco** [comment]
@tseaver I've never noticed that before!  I really didn't expect it in a "HACKING" file.  It would be great if it were somehow pointed to in the README and highlighted at the top of that file (ie, a TOC of what's in that).   An example for the individual test would be great too.  I'm really biased to this sort of stuff being "super obvious".

**Comment by stevepiercy** [comment]
See also https://github.com/Pylons/pyramid/blob/master/contributing.md#prerequisites which appears as a [friendly yellow reminder](https://help.github.com/articles/setting-guidelines-for-repository-contributors/) whenever you report an issue or submit a PR.

All that said, I agree with @jvanasco. "HACKING" is not an intuitive name. I recall I had the same problem when I was trying to get started with Pyramid. I usually check the README first. I'll add a section to the README, and roll it down to 1.5.

**Comment by jvanasco** [comment]
@stevepiercy maybe something like this in the README ?

	Running Tests and Development
	-----------------------------

	See the `HACKING.rst` file to learn how Pyramid runs tests and developer tips 
	and requirements when contributing to this project.

**Comment by bertjwregeer** [comment]
As a developer on a huge variety of projects over the years. `HACKING` is very often used as the file name for developers that want to "hack" on the project. I don't find it that unintuitive.

**Comment by marioidival** [comment]
Guys, I never used Windows and Python together, you can see if this snippet is correct?
````sh
    c:\\> c:\\Python33\\python -m venv env33
    c:\\> env33\\Scripts\\activate
    c:\\> pip install "pyramid==\ |release|\ "
```

Snippet in `pyramid/docs/quick_tour.rst`

**Comment by mmerickel** [comment]
@marioidival please don't hijack this ticket, it isn't the appropriate place for support questions (the mailing list is). You can see in http://docs.pylonsproject.org/projects/pyramid/en/1.6-branch/quick_tour.html that the RST source is rendered out as `pyramid==1.6`, not `"pyramid==\ |release|\ "`.

**Comment by stevepiercy** [comment]
@marioidival it renders as:

```bash
c:\> c:\Python33\python -m venv env33
c:\> env33\Scripts\python ez_setup.py
c:\> env33\Scripts\easy_install "pyramid==1.6"
```

See rendered result: http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tour.html#installation

See also the rst directive [parsed-literal](http://docutils.sourceforge.net/docs/ref/rst/directives.html#parsed-literal).

**Comment by marioidival** [comment]
@mmerickel Here is the correct place so, why am I doing this PR, started programming using UNIX environments and as I said before, never used Windows, I wonder how to do so correctly to proceed.

**Comment by Themanwithoutaplan** [comment]
1) tox + requirements.txt inflating setup.py with optional dependencies makes things unnecessarily complicated.

**Comment by stevepiercy** [comment]
@mmerickel @dstufft's [article](https://caremad.io/2013/07/setup-vs-requirement/) suggests a fourth option (NB: I am not a package manager, nor do I play one on television.)

```ini
# requirements.txt
--index-url https://pypi.python.org/pyramid/

-e .
```

which would be invoked by `pip install -r requirements.txt`. At the same time, remove the corresponding stanzas from `setup.py` (I know, sacrilege, right?), specifically https://github.com/Pylons/pyramid/blob/master/setup.py#L41-L50 and https://github.com/Pylons/pyramid/blob/master/setup.py#L103

That pattern could be repeated for `tests-requirements.txt`, `docs-requirements.txt`, and any other purpose.

I think that's what @Themanwithoutaplan meant.

**Comment by Themanwithoutaplan** [comment]
setup.py must still have the packages required for the library. Outside of a testing_requirements.txt which is useful for a test-runner without invoking tox (a very common pattern), I wouldn't want any additional requirements files because everything can be driven by tox, unless tox uses them too. This is a clearer separation of responsibilities.

**Comment by bertjwregeer** [comment]
What? I hate requirements.txt, I hate the fact that it means now I need to run `pip install -e .` and then `pip install -r requirements.txt` to get started on a project.

`pip install -e .[testing]` also makes sense to me. Keep everything in one place, don't try and "split responsibilities".

**Comment by stevepiercy** [comment]
Whoa, I don't know what I did, but somehow this issue got closed. Reopening.

The branch for this work is now https://github.com/Pylons/pyramid/tree/docs/easy-install-to-pip.2104

**Comment by stevepiercy** [comment]
While working on the wiki2 tutorial, I edited the default `setup.py` file in order to install test requirements as @mmerickel described in Option 3.

1. Should I roll changes in `setup.py` into the alchemy scaffold or just leave them in the wiki2 tutorial?
1. While I'm at it—and because the wiki2 tutorial ends up using this structure anyway—should I use the [Conventions for Python test discovery](http://pytest.org/latest/goodpractices.html#test-discovery), i.e., 

```text
tests/
    __init__.py
    test_default.py
```

I'm leaning toward rolling these changes into the scaffold so that it provides a good example of how to get started.

Let me know what y'all think.

**Comment by Themanwithoutaplan** [comment]
-1 on testing requirements in setup.py for the usual reasons: a tox config is much more flexible and in any case required for serious testing. 

If you're going to following the pytest recommendations then add a pytest.ini (norecursedirs is a life-saver) as this makes the test discovery pattern explicit.

**Comment by stevepiercy** [comment]
@Themanwithoutaplan I scratched the pytest recommendations because some people like to group tests in a single file, others by folder, and others by feature. Better not to exert an opinion on this one.

As far as tox or setup.py is used, I have no opinion other than going with the prevailing opinion, which is currently Option 3, so that's what I'm documenting.

Of course, if the wind should change, I can go with tox.

**Comment by stevepiercy** [comment]
See PR #2442 for the rough draft of `setup.py` for Option 3.

**Comment by jvanasco** [comment]
I just created a project using that forthcoming scaffold. 
The tests starter would have been great.

**Comment by bertjwregeer** [comment]
@Themanwithoutaplan we currently don't provide a tox.ini in our projects, and I don't think we should add that as yet another thing in the scaffold.

I am a +1 on option 3, adding testing extra's makes sense, and is a built-in functionality that exists specifically for adding extra dependencies.

**Comment by bertjwregeer** [comment]
@stevepiercy I do think you should roll them back into the scaffold.

**Comment by tseaver** [comment]
`tox` is hugely important for library developers, who need to ensure that their tests run easily on all supported Python versions / implementations.  I'm not convinced that `tox` is a huge win for the kind of apps we expect people to build with the scaffolds:  "bespoke" apps tend to be deployed only for a single Python version, for instance, which negates the main advantage of `tox` over other testing styles.

**Comment by pauleveritt** [comment]
I'm also -1 on tox. Scaffolds are (primarily) for new folks. Handing them another thing to learn and deal with should be kept to a minimum. In JS/Yeoman, a lot of scaffolds have a dozen flavor-of-the-week JS project/tools in them, and the scaffold winds up being brittle magic.

**Comment by Themanwithoutaplan** [comment]
Even though I introduced it to the discussion, debate about tox is a bit of a red herring. To mix my food metaphors: the old issues  of a testing clause versus testing_requirements.txt is really about what kind of cheese people are used to. As I **still** struggle with writing setup.py but can reliably produce and manage requirements files this is why I'm in favour of them. Adding options to setup.py adds complexity which I think the incantation exposes: try and explain that to anyone new.

But I'm happy to be in my minority of one. I usually am. :-)

**Comment by stevepiercy** [comment]
I've hit a roadblock on the Installation chapter: installation of [setuptools](http://docs.pylonsproject.org/projects/pyramid/en/master/narr/install.html#installing-setuptools) and [virtualenv](http://docs.pylonsproject.org/projects/pyramid/en/master/narr/install.html#installing-the-virtualenv-package). I could use some clarity from the experienced folks.

I'm not sure how to rewrite these two sections. I'm inclined to prioritize Python 3.4+, with exceptions to the priority.

* For Python 3.4+, both setuptools and virtualenv are already installed.
* For Python 3.3... ?
* For Python 2.7... ?
* For Python 2.6... ?

**Comment by stevepiercy** [comment]
A *huge, major, ginormous* change for Installation of Pyramid is now in PR #2466. I'm going to let this one simmer for a while for discussion.

**Comment by stevepiercy** [comment]
I'm DONE! whew! See #2468.

**Comment by stevepiercy** [comment]
Closed by #2468
